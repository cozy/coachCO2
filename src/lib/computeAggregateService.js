import difference from 'lodash/difference'
import { CCO2_SETTINGS_DOCTYPE, JOBS_DOCTYPE } from 'src/doctypes'
import { buildAggregateCaptureDeviceTimeseriesQuery } from 'src/queries/nodeQueries'
import { buildSettingsQuery } from 'src/queries/queries'

import logger from 'cozy-logger'

const logService = logger.namespace('services/computeAggregateService')

/**
 * @param {import('cozy-client/types/CozyClient').default} client - The cozy client
 */
export const runComputeAggregateService = async client => {
  const settings = await client.queryAll(buildSettingsQuery().definition)
  if (!settings || settings?.length < 1) {
    logService('error', 'App settings not found')
    return false
  }
  const settingsCaptureDevices = settings[0].captureDevices || []

  const timeseries = await client.queryAll(
    buildAggregateCaptureDeviceTimeseriesQuery().definition
  )
  const timeseriesCaptureDevices = [
    ...new Set(timeseries.map(ts => ts.captureDevice))
  ]

  const isAlreadyUpdated =
    settingsCaptureDevices.length === timeseriesCaptureDevices.length &&
    difference(settingsCaptureDevices, timeseriesCaptureDevices).length === 0

  if (isAlreadyUpdated) {
    logService('info', 'captureDevices in app settings are already up to date')
    return false
  }

  await client.save({
    ...settings[0],
    _type: CCO2_SETTINGS_DOCTYPE,
    captureDevices: timeseriesCaptureDevices
  })

  logService('info', 'captureDevices in app settings updated successfully')
}

/**
 * @param {import('cozy-client/types/CozyClient').default} client
 */
export const launchComputeAggregateJob = async client => {
  try {
    logService('info', 'Start launchMetadataMigrationJob')
    const settings = await client.queryAll(buildSettingsQuery().definition)
    const setting = settings?.[0] || {}

    if (setting?.captureDevices) {
      logService(
        'info',
        'Stop launchComputeAggregateJob because captureDevices is already exists in app settings'
      )
      return
    }

    logService(
      'info',
      "Create job service with slug: 'coachco2' & name: 'computeAggregate'"
    )
    const jobColl = client.collection(JOBS_DOCTYPE)
    await jobColl.create(
      'service',
      { slug: 'coachco2', name: 'computeAggregate' },
      {},
      true
    )
  } catch (error) {
    logService('error', `launchComputeAggregateJob error: ${error}`)
  }
}
