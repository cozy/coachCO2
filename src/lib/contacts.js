import { COORDINATES_DISTANCE_THRESHOLD_M } from 'src/constants'
import {
  getEndPlaceCoordinates,
  getStartPlaceCoordinates
} from 'src/lib/timeseries'

import { geodesicDistance } from 'cozy-client/dist/models/geo'

/**
 * @typedef {import('./types').TimeseriesGeoJSON} TimeSerie
 * @typedef {import('./types').Contact} Contact
 * @typedef {import('./types').Address} Address
 */

/**
 * Find the closest addresses related to a start and end place
 *
 * @typedef StartAndEndClosestAddress
 * @property {Address} matchingStart
 * @property {Address} matchingEnd
 *
 * @param {TimeSerie} timeserie - The timeserie with start and end place coordinates
 * @param {Array<Contact>} contacts - The list of contacts to search for an address
 * @returns {StartAndEndClosestAddress} The closest addresses for start and end place
 */
export const findClosestStartAndEndContactsAddress = (timeserie, contacts) => {
  const startCoordinates = getStartPlaceCoordinates(timeserie)
  const endCoordinates = getEndPlaceCoordinates(timeserie)
  let closestStart = {},
    closestEnd = {}
  let minStartDistance = Number.MAX_SAFE_INTEGER,
    minEndDistance = Number.MAX_SAFE_INTEGER
  for (const contact of contacts) {
    for (const address of contact.address) {
      if (!address?.geo?.geo) {
        continue
      }
      const addressCoordinates = {
        lon: address.geo.geo[0],
        lat: address.geo.geo[1]
      }
      const startPlaceDistance = geodesicDistance(
        startCoordinates,
        addressCoordinates
      )

      const endPlaceDistance = geodesicDistance(
        endCoordinates,
        addressCoordinates
      )

      if (startPlaceDistance < minStartDistance) {
        closestStart = {
          distance: startPlaceDistance,
          contact,
          address,
          newCoordinates: startCoordinates
        }
        minStartDistance = startPlaceDistance
      }
      if (endPlaceDistance < minEndDistance) {
        closestEnd = {
          distance: endPlaceDistance,
          contact,
          address,
          newCoordinates: endCoordinates
        }
        minEndDistance = endPlaceDistance
      }
    }
  }
  return { closestStart, closestEnd }
}

/**
 *
 * For a timeserie, find a matching contact address for start and end place.
 * A matching address has the closest coordinates and is below a distance threshold.
 *
 * @typedef StartAndEndMatchingAddress
 * @property {Address} matchingStart
 * @property {Address} matchingEnd
 *
 * @param {TimeSerie} timeserie - The timeserie with start and end place coordinates
 * @param {Array<Contact>} contacts - The list of contacts to search for an address
 * @returns {StartAndEndMatchingAddress} The matching start and end addresses
 */
export const findMatchingStartAndEnd = (timeserie, contacts) => {
  const { closestStart, closestEnd } = findClosestStartAndEndContactsAddress(
    timeserie,
    contacts
  )
  let matchingStart = null,
    matchingEnd = null
  if (closestStart?.distance < COORDINATES_DISTANCE_THRESHOLD_M) {
    matchingStart = closestStart
  }
  if (closestEnd?.distance < COORDINATES_DISTANCE_THRESHOLD_M) {
    matchingEnd = closestEnd
  }
  return { matchingStart, matchingEnd }
}

/**
 *
 * Try to find a contact address matching the given timeserie start or end place.
 *
 * @typedef {Object} params
 * @property {TimeSerie} - The timeserie to extract start and end coordinates
 * @property {Contact} - The contact containing the address to find a matching
 * @property {startOrEnd} - Either 'start' or 'end'. Specifiy if it should find a matching on the trip start or end
 * @param Params
 * @returns { Address} - The matching contact address
 */
export const findMatchingContactAddressForTimeserie = ({
  timeserie,
  contact,
  startOrEnd
}) => {
  const { matchingStart, matchingEnd } = findMatchingStartAndEnd(timeserie, [
    contact
  ])
  if (startOrEnd === 'start') {
    return matchingStart?.address || null
  } else if (startOrEnd === 'end') {
    return matchingEnd?.address || null
  }
  return null
}
