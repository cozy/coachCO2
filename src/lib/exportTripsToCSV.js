import sortBy from 'lodash/sortBy'

import {
  getStartPlaceDisplayName,
  getEndPlaceDisplayName,
  getPurpose,
  getSectionsInfo,
  getStartDate,
  getEndDate
} from 'src/lib/trips'
import { COLUMNS_NAMES_CSV } from 'src/constants/const'

/**
 * @param {object} tripsData
 * @returns {string}
 */
const convertTripsToCSV = tripsData => {
  const tripsDataWithHeader = [Object.keys(tripsData[0]), ...tripsData]

  return tripsDataWithHeader
    .map(trip =>
      Object.values(trip)
        .map(tripValue => (tripValue ? `"${tripValue}"` : ' '))
        .toString()
    )
    .join('\n')
}

/**
 * @param {object} trips
 * @returns {object}
 */
export const makeTripsForExport = trips => {
  const result = trips.flatMap(trip => {
    const sectionsInfo = getSectionsInfo(trip)

    return sectionsInfo.map(sectionInfo => {
      const sectionData = {
        [COLUMNS_NAMES_CSV.sectionStartDate]: sectionInfo.startDate,
        [COLUMNS_NAMES_CSV.sectionEndDate]: sectionInfo.endDate,
        [COLUMNS_NAMES_CSV.sectionDuration]: sectionInfo.duration,
        [COLUMNS_NAMES_CSV.sectionDistance]: sectionInfo.distance,
        [COLUMNS_NAMES_CSV.sectionMode]: sectionInfo.mode,
        [COLUMNS_NAMES_CSV.sectionCoordinates]: sectionInfo.coordinates,
        [COLUMNS_NAMES_CSV.sectionDistances]: sectionInfo.distances,
        [COLUMNS_NAMES_CSV.sectionTimestamps]: sectionInfo.timestamps,
        [COLUMNS_NAMES_CSV.sectionSpeeds]: sectionInfo.speeds
      }
      const tripData = {
        [COLUMNS_NAMES_CSV.tripId]: trip.id,
        [COLUMNS_NAMES_CSV.tripStartDate]: getStartDate(trip),
        [COLUMNS_NAMES_CSV.tripEndDate]: getEndDate(trip),
        [COLUMNS_NAMES_CSV.tripStartDisplayName]: getStartPlaceDisplayName(
          trip
        ),
        [COLUMNS_NAMES_CSV.tripArrivalDisplayName]: getEndPlaceDisplayName(
          trip
        ),
        [COLUMNS_NAMES_CSV.tripPurpose]: getPurpose(trip)
      }

      return { ...tripData, ...sectionData }
    })
  })

  return sortBy(result, COLUMNS_NAMES_CSV.tripId)
}

/**
 * @param {object} trips
 * @returns {string}
 */
export const exportTripsToCSV = trips => {
  if (!trips || trips.length === 0) return

  const tripsData = makeTripsForExport(trips)

  if (tripsData.length > 0) {
    return convertTripsToCSV(tripsData)
  }
}
