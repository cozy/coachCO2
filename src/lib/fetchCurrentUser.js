import { CONTACTS_DOCTYPE } from 'src/doctypes'

/**
 * Fetch current user
 * @param {import('cozy-client/types/CozyClient').default} client - CozyClient instance
 * @returns {Promise<Object>} - Current user
 */
export const fetchCurrentUser = async client => {
  const contactCollection = client.collection(CONTACTS_DOCTYPE)
  const { data: currentUser } = await contactCollection.findMyself()

  return currentUser[0]
}
