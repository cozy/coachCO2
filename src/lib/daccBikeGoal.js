import format from 'date-fns/format'
import subDays from 'date-fns/subDays'
import { getSource } from 'src/components/Goals/BikeGoal/helpers'
import { DACC_MEASURE_NAME_BIKE_GOAL } from 'src/constants'
import {
  createMeasureForDACC,
  getDACCRemoteDoctype,
  sendMeasureToDACCWithRemoteDoctype
} from 'src/lib/dacc'
import { countUniqDays } from 'src/lib/timeseries'
import { buildBikeCommuteTimeseriesQueryByAccountId } from 'src/queries/queries'

import { models } from 'cozy-client'
import flag from 'cozy-flags'
import log from 'cozy-logger'

const { fetchAggregatesFromDACC } = models.dacc

/**
 * @typedef {import("cozy-client/dist/index").CozyClient} CozyClient
 * @typedef {import("cozy-client/types/types").IoCozyAccount} IoCozyAccount
 * @typedef {import('cozy-client/types/types').DACCAggregate} DACCAggregate
 */

/**
 * Fetch bike goal aggregate, for the previous day
 *
 * @param {CozyClient} client - The cozy-client instance
 * @returns {Array<DACCAggregate>} The aggregates sorted by startDate.
 */
export const fetchYesterdayBikeGoalFromDACC = async client => {
  try {
    const remoteDoctype = getDACCRemoteDoctype()
    // Get results from previous day, to ensure there is data
    const currentDate = Date.now()
    const startDate = format(subDays(currentDate, 1), 'yyyy-MM-dd')
    const endDate = format(currentDate, 'yyyy-MM-dd')
    const results = await fetchAggregatesFromDACC(client, remoteDoctype, {
      measureName: DACC_MEASURE_NAME_BIKE_GOAL,
      startDate,
      endDate
    })
    return results
  } catch (error) {
    log(
      'error',
      `Error while retrieving data from remote-doctype: ${error.message}`
    )
  }
}

/**
 * Get average value for the given group name
 *
 * @param {Array<DACCAggregate>} aggregates - The aggregates to extract value
 * @param {string} groupName - The group name from which we want the average value
 * @returns {number} The average value for the group
 */
export const getAvgDaysForGroupName = aggregates => {
  const groupName = getBikeGroupName()
  for (const agg of aggregates) {
    const isCorrectGroup = agg.groups.some(
      group => group.groupName === groupName
    )
    if (isCorrectGroup) {
      return Math.round(agg.avg)
    }
  }
  return null
}

/**
 * The group name provided by flag
 *
 * @param {CozyClient}} client - The cozy-client instance
 * @returns {string} The group name
 */
export const getBikeGroupName = () => {
  const { sourceName } = getSource()
  const groupName = sourceName || 'unknown'
  return groupName
}

/**
 * Send bike goal measures
 *
 * @param {CozyClient}} client - The cozy-client instance
 * @param {IoCozyAccount} account - The trips account to measure
 */
export const sendBikeGoalMeasuresForAccount = async (client, account) => {
  const currentDate = new Date()
  await flag.initialize(client)

  const bikeCommuteQuery = buildBikeCommuteTimeseriesQueryByAccountId(
    { accountId: account._id, date: currentDate },
    true
  ).definition
  log('info', `Query commute trips with: ${JSON.stringify(bikeCommuteQuery)}`)
  const timeseries = await client.queryAll(bikeCommuteQuery)
  log('info', `Number of commute trips on bike: ${timeseries?.length}`)
  if (timeseries?.length < 1) {
    log('info', 'No timeseries found. Abort.')
    return null
  }

  const daysWithBikeCommute = countUniqDays(timeseries)
  log('info', `Number of days with commute bike : ${daysWithBikeCommute}`)

  const bikeGroupName = getBikeGroupName()

  const measure = createMeasureForDACC({
    startDate: currentDate,
    value: daysWithBikeCommute,
    measureName: DACC_MEASURE_NAME_BIKE_GOAL,
    groups: [
      {
        groupName: bikeGroupName
      }
    ]
  })
  log('info', `Send measure ${JSON.stringify(measure)} to DACC...`)
  await sendMeasureToDACCWithRemoteDoctype(client, measure)
}
