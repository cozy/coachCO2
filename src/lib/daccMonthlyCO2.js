import addMonths from 'date-fns/addMonths'
import format from 'date-fns/format'
import isThisMonth from 'date-fns/isThisMonth'
import startOfMonth from 'date-fns/startOfMonth'
import subMonths from 'date-fns/subMonths'
import {
  DACC_MEASURE_NAME_CO2_MONTHLY,
  DACC_MEASURE_GROUP1_CO2_MONTHLY,
  TIMESERIE_MIGRATION_SERVICE_NAME,
  MAX_DACC_MEASURES_SENT
} from 'src/constants'
import {
  createMeasureForDACC,
  getDACCRemoteDoctype,
  sendMeasureToDACCWithRemoteDoctype
} from 'src/lib/dacc'
import { startService } from 'src/lib/services'
import { computeCO2Timeseries } from 'src/lib/timeseries'
import {
  buildTimeseriesQueryByDateAndAccountId,
  buildOldestTimeseriesQueryByAccountId
} from 'src/queries/queries'

import { models } from 'cozy-client'
import log from 'cozy-logger'

const { fetchAggregatesFromDACC } = models.dacc

/**
 * @typedef {object} DACCAggregate
 *
 * See https://github.com/cozy/cozy-client/blob/0724ceb0905923fc2b2f2e7f322498042813a3b8/packages/cozy-client/src/types.js#L340
 *
 */
/**
 * Fetch monthly DACC CO2 aggregates
 *
 * @param {object} client - The cozy-client instance
 * @returns {Array<DACCAggregate>} The aggregates sorted by startDate.
 */
export const fetchMonthlyAverageCO2FromDACCFor11Month = async client => {
  try {
    const remoteDoctype = getDACCRemoteDoctype()
    const startDate = format(
      startOfMonth(subMonths(Date.now(), 11)),
      'yyyy-MM-dd'
    )
    const results = await fetchAggregatesFromDACC(client, remoteDoctype, {
      measureName: DACC_MEASURE_NAME_CO2_MONTHLY,
      startDate
    })
    return results
  } catch (error) {
    log(
      'error',
      `Error while retrieving data from remote-doctype: ${error.message}`
    )
  }
}

const fetchTimeseriesForMonth = async (client, accountId, startDate) => {
  const query = buildTimeseriesQueryByDateAndAccountId(startDate, accountId, {
    limit: 1000
  }).definition
  return client.queryAll(query)
}

export const hasNonAggregatedTimeseries = timeseries => {
  const timeseriesWithAggregation = timeseries.filter(
    timeserie => timeserie.aggregation
  )
  return timeseriesWithAggregation.length < timeseries.length
}

const getNextMonthStartDate = date => {
  return startOfMonth(addMonths(date, 1))
}

export const getNextMeasureStartDate = date => {
  const nextMonthStartDate = getNextMonthStartDate(date)
  if (
    nextMonthStartDate.getTime() > Date.now() ||
    isThisMonth(nextMonthStartDate)
  ) {
    return null
  }
  return nextMonthStartDate
}

export const getStartDate = async (client, account) => {
  const accountDate = account.data?.lastDACCMeasureStartDate
  const lastDACCMeasureStartDate = accountDate ? new Date(accountDate) : null
  if (lastDACCMeasureStartDate) {
    return new Date(lastDACCMeasureStartDate)
  }

  const query = buildOldestTimeseriesQueryByAccountId(account._id).definition
  const oldestDoc = await client.query(query)
  if (!oldestDoc.data || oldestDoc.data.length < 1) {
    return null
  }
  const startDate = new Date(oldestDoc.data[0].startDate)

  return startOfMonth(subMonths(startDate, 1))
}

const saveStartDateInAccount = async (client, account, date) => {
  const newAccount = {
    ...account,
    data: {
      ...account.data,
      lastDACCMeasureStartDate: date
    }
  }
  await client.save(newAccount)
}

const createCO2MeasureToDACC = (startDate, value) => {
  return createMeasureForDACC({
    startDate,
    value,
    measureName: DACC_MEASURE_NAME_CO2_MONTHLY,
    group1: DACC_MEASURE_GROUP1_CO2_MONTHLY
  })
}

export const sendCO2MeasuresForAccount = async (client, account) => {
  let nMeasuresSent = 0
  let startDate = await getStartDate(client, account)
  if (!startDate) {
    log('info', `No data to process for account ${account._id}.`)
    return nMeasuresSent
  }
  log('info', `Start measures from ${startDate.toISOString()}`)
  let nextStartDate = getNextMeasureStartDate(startDate)
  while (nextStartDate && nMeasuresSent < MAX_DACC_MEASURES_SENT) {
    startDate = nextStartDate
    const timeseries = await fetchTimeseriesForMonth(
      client,
      account._id,
      nextStartDate
    )
    if (timeseries.length < 1) {
      // No trips to process for this month
      log('info', `No trips for ${startDate.toISOString()}`)
      nextStartDate = getNextMeasureStartDate(startDate)
      continue
    }
    if (hasNonAggregatedTimeseries(timeseries)) {
      // There are timeseries without aggregation: interrupt service execution to run migration service
      log(
        'warn',
        `Timeseries for ${startDate.toISOString()} does not have aggregation`
      )
      await startService(client, TIMESERIE_MIGRATION_SERVICE_NAME)
      log('info', 'Timeseries migration service started')
      return nMeasuresSent
    }
    const CO2 = computeCO2Timeseries(timeseries)
    const measure = createCO2MeasureToDACC(startDate, CO2)
    await sendMeasureToDACCWithRemoteDoctype(client, measure)
    nMeasuresSent++
    nextStartDate = getNextMeasureStartDate(startDate)
  }
  await saveStartDateInAccount(client, account, startDate)
  log('info', `Saved  ${startDate.toISOString()} in account`)

  return nMeasuresSent
}
