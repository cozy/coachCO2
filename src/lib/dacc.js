import { MAX_DACC_MEASURES_SENT } from 'src/constants'
import { DACC_REMOTE_DOCTYPE, DACC_REMOTE_DOCTYPE_DEV } from 'src/doctypes'
import { buildAccountQuery, buildSettingsQuery } from 'src/queries/queries'

import { models } from 'cozy-client'
import flag from 'cozy-flags'
import log from 'cozy-logger'

import { sendMeasuresForAccount } from './daccMonthlyCO2'

const { sendMeasureToDACC } = models.dacc

export const getDACCRemoteDoctype = () => {
  return flag('coachco2.dacc-dev_v2') === true
    ? DACC_REMOTE_DOCTYPE_DEV
    : DACC_REMOTE_DOCTYPE
}

export const sendMeasureToDACCWithRemoteDoctype = async (client, measure) => {
  try {
    await flag.initialize(client)
    const remoteDoctype = getDACCRemoteDoctype()
    log('info', `Send measure to ${remoteDoctype} on ${measure.startDate}`)

    await sendMeasureToDACC(client, remoteDoctype, measure)
  } catch (error) {
    log(
      'error',
      `Error while sending measure to remote doctype: ${error.message}`
    )
    throw error
  }
}

/**
 * Run the DACC service
 *
 * @param {object} client - The client instance
 * @returns {boolean} Whether or not the service should be restarted
 */
export const runDACCService = async client => {
  const settings = await client.queryAll(buildSettingsQuery().definition)
  if (!settings?.[0]?.CO2Emission?.sendToDACC) {
    log('info', 'The user did not give consent to send data to DACC')
    return false
  }
  const accounts = await client.queryAll(
    buildAccountQuery({ limit: 1000, withOnlyLogin: false }).definition
  )
  if (!accounts) {
    log('info', 'No account found: Nothing to do')
    return false
  }

  let shouldRestartService = false

  for (const account of accounts) {
    const nMeasuresSent = await sendMeasuresForAccount(client, account)
    if (nMeasuresSent >= MAX_DACC_MEASURES_SENT) {
      shouldRestartService = true
    }
    log(
      'info',
      `${nMeasuresSent} measures sent to DACC for account ${account._id}`
    )
  }
  return shouldRestartService
}
