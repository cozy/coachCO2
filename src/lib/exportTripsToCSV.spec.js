import {
  mockSerie,
  mockFeature,
  mockFeatureCollection,
  modeProps
} from 'test/mockTrip'

import { exportTripsToCSV } from 'src/lib/exportTripsToCSV'

jest.mock('src/lib/exportTripsToCSV', () => ({
  ...jest.requireActual('src/lib/exportTripsToCSV'),
  makeTripsForExport: jest.fn()
}))

const mockedFeatures = () => [
  mockFeature('featureId01'),
  mockFeature('featureId02'),
  mockFeature('featureId03'),
  mockFeature('featureId04'),
  mockFeatureCollection('featureCollectionId01', [
    mockFeature('featureId05', modeProps.bicycle)
  ]),
  mockFeatureCollection('featureCollectionId02', [
    mockFeature('featureId06', modeProps.walking)
  ]),
  mockFeatureCollection('featureCollectionId03', [
    mockFeature('featureId07', modeProps.car)
  ])
]

describe('exportTripsToCSV', () => {
  const trips = mockSerie('serieId01', mockedFeatures())

  it('should return correctly formatted trips for the CSV file', () => {
    const tripCSV = exportTripsToCSV([trips])

    expect(tripCSV).toMatchSnapshot()
  })

  it('should return undefined', () => {
    const tripCSV = exportTripsToCSV()

    expect(tripCSV).toBeUndefined()
  })
})
