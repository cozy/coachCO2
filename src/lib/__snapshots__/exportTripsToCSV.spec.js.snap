// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exportTripsToCSV should return correctly formatted trips for the CSV file 1`] = `
"\\"Trip_Id\\",\\"Trip_Start_Date\\",\\"Trip_End_Date\\",\\"Trip_Start_Display_Name\\",\\"Trip_Arrival_Display_Name\\",\\"Trip_Purpose\\",\\"Section_Start_Date\\",\\"Section_End_Date\\",\\"Section_Duration\\",\\"Section_Distance\\",\\"Section_Mode\\",\\"Section_Coordinates\\",\\"Section_Distances\\",\\"Section_Timestamps\\",\\"Section_Speeds\\"
\\"serieId01\\",\\"Wed Jun 30 2021 14:47:51 GMT+0200 (heure d’été d’Europe centrale)\\",\\"Wed Jun 30 2021 16:37:05 GMT+0200 (heure d’été d’Europe centrale)\\",\\"GR9, Isère\\",\\"Piste de la Combe Noire, Isère\\",\\"PICK_DROP\\",\\"Tue Dec 07 2021 16:24:49 GMT+0100 (heure normale d’Europe centrale)\\",\\"Tue Dec 07 2021 16:39:16 GMT+0100 (heure normale d’Europe centrale)\\",\\"600\\",\\"2456\\",\\"BICYCLING\\",\\"2.31251,48.7799432,2.313591181324194,48.78161416932299,2.314672362648388,48.783285138645994\\",\\"0,201.98652472578271,201.98549037826737\\",\\"1638890689000,1638890719000,1638890749000\\",\\"0,6.73288415752609,6.732849679275579\\"
\\"serieId01\\",\\"Wed Jun 30 2021 14:47:51 GMT+0200 (heure d’été d’Europe centrale)\\",\\"Wed Jun 30 2021 16:37:05 GMT+0200 (heure d’été d’Europe centrale)\\",\\"GR9, Isère\\",\\"Piste de la Combe Noire, Isère\\",\\"PICK_DROP\\",\\"Tue Dec 07 2021 16:39:31 GMT+0100 (heure normale d’Europe centrale)\\",\\"Tue Dec 07 2021 17:04:07 GMT+0100 (heure normale d’Europe centrale)\\",\\"540\\",\\"563\\",\\"WALKING\\",\\"2.31251,48.7799432,2.313591181324194,48.78161416932299,2.314672362648388,48.783285138645994\\",\\"0,2.3338746642432437,5.89018714479141\\",\\"1638891571000,1638891601000,1638891631000\\",\\"0,0.07779582214144146,0.196339571493047\\"
\\"serieId01\\",\\"Wed Jun 30 2021 14:47:51 GMT+0200 (heure d’été d’Europe centrale)\\",\\"Wed Jun 30 2021 16:37:05 GMT+0200 (heure d’été d’Europe centrale)\\",\\"GR9, Isère\\",\\"Piste de la Combe Noire, Isère\\",\\"PICK_DROP\\",\\"Tue Dec 07 2021 17:04:34 GMT+0100 (heure normale d’Europe centrale)\\",\\"Tue Dec 07 2021 17:06:26 GMT+0100 (heure normale d’Europe centrale)\\",\\"1800\\",\\"14789\\",\\"CAR\\",\\"2.31251,48.7799432,2.313591181324194,48.78161416932299,2.314672362648388,48.783285138645994\\",\\"0,43.094121726730805,31.5918459191552\\",\\"1638893074000,1638893104000,1638893134000\\",\\"0,1.4364707242243602,1.0530615306385067\\""
`;
